### PR TITLE
Feature/corutines network layer

### PR DESCRIPTION
**Pull request by amaury901130**
### Description
- Proposal coroutines on the network layer, over the ViewModelScope

Official doc: 
https://developer.android.com/topic/libraries/architecture/coroutines

Issue: [Add Kotlin Coroutines](https://github.com/rootstrap/android-base/issues/22)

### CODE REVIEW

File: app/src/main/java/com/rootstrap/android/ui/activity/main/SignInActivityViewModel.kt [hunk 1]
```diff
@@ -23,26 +26,28 @@ open class SignInActivityViewModel(listener: ViewModelListener?) : BaseViewModel
...
+        }
     }
 
-    @Subscribe
-    fun signedInError(event: ErrorEvent) {
-        error = event.error
-        networkState = NetworkState.idle
-        networkState = NetworkState.error
-    }
+    private fun mangeError(exception: Throwable?) {
```
Comment (ximenaperez) [inline app/src/main/java/com/rootstrap/android/ui/activity/main/SignInActivityViewModel.kt:44]: there's a typo in the method name, it should be manageError

File: app/src/main/java/com/rootstrap/android/ui/activity/main/SignUpActivityViewModel.kt [hunk 2]
```diff
@@ -23,26 +26,29 @@ open class SignUpActivityViewModel(listener: ViewModelListener?) : BaseViewModel
...
+            } else {
+                mangeError(result.exceptionOrNull())
+            }
+        }
     }
 
-    @Subscribe
-    fun signedUpFailure(event: FailureEvent) {
-        error = null
+    private fun mangeError(exception: Throwable?) {
```
Comment (ximenaperez) [inline app/src/main/java/com/rootstrap/android/ui/activity/main/SignUpActivityViewModel.kt:45]: same here

File: app/src/main/java/com/rootstrap/android/util/extensions/ActionCallback.kt [hunk 3]
```diff
@@ -1,44 +1,57 @@
...
+) : java.lang.Exception() {
+    override val message: String?
+        get() = errorMessage
 }
 
-class ErrorEvent(val error: String)
-class FailureEvent
+enum class ApiErrorType {
+    apiError,
+    uknownError
```
Comment (ximenaperez) [inline app/src/main/java/com/rootstrap/android/util/extensions/ActionCallback.kt:56]: there's a typo here, it should be "unknown"

**Review by ximenaperez — CHANGES_REQUESTED** [review]
Just a few typos to fix

**Review by ximenaperez — APPROVED** [review]

File: app/src/main/java/com/rootstrap/android/util/extensions/ActionCallback.kt [hunk 4]
```diff
@@ -1,44 +1,57 @@
...
+                val response = apiCall.execute()
+                manageResponse(response)
+            }
 
-    open fun errorAction(response: Response<T>) {
-        try {
-            response.errorBody()?.let {
-                val error = Gson().fromJson(it.charStream(), ErrorModel::class.java)
-                bus.post(ErrorEvent(ErrorUtil.handleCustomError(error)))
+        private fun <T> manageResponse(response: Response<T>): Result<Data<T>> {
```
Comment (mato2593) [inline app/src/main/java/com/rootstrap/android/util/extensions/ActionCallback.kt:20]: IMO, `handleResponse` is a more accurate name of what we are doing here. The same would apply for the other `manageSomething` functions.

File: app/src/main/java/com/rootstrap/android/ui/activity/main/SignUpActivityViewModel.kt [hunk 5]
```diff
@@ -23,26 +26,29 @@ open class SignUpActivityViewModel(listener: ViewModelListener?) : BaseViewModel
...
+                    SessionManager.signIn(user)
+                }
 
-    @Subscribe
-    fun signedUpError(event: ErrorEvent) {
-        error = event.error
-        networkState = NetworkState.idle
-        networkState = NetworkState.error
+                networkState = NetworkState.idle
+                state = SignUpState.signedUpSuccess
```
Comment (mato2593) [inline app/src/main/java/com/rootstrap/android/ui/activity/main/SignUpActivityViewModel.kt:38]: I know this is old, but it should be either `signUpSuccess` or `signedUpSuccessfully`.

**Review by mato2593 — APPROVED** [review]

**Pull request merged**